### PR TITLE
Replace @requiresTypeResolution kdoc tag with @RequiresTypeResolution annotation

### PR DESCRIPTION
**Pull request by marschwar**
This PR belongs to #3562 and replaces the @requiresTypeResolution kdoc tag with a @RequiresTypeResolution annotation.

There are a few things that I would like to point out/discuss:
* I still have some formatting issues. That is why there is one commit that just reformats the code according to the ktlint plugin
* I added the annotation to the detekt API which might be a bit premature. Maybe it should be placed in `internal` until it is stable?
* I started out adding `@UnstableApi` to the annotation but that would all rules to require to add `@OptIn(UnstableApi::class)` which does not feel right either.
* Adding a `@Suppress("TooManyFunctions")` to the `RuleVisitor` does not feel great, but I thinkt the code reads much better with those extension functions. Suggestions are welcome.

### CODE REVIEW

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleVisitor.kt [hunk 1]
```diff
@@ -17,7 +18,9 @@ import org.jetbrains.kotlin.psi.KtValueArgument
 import org.jetbrains.kotlin.psi.psiUtil.containingClass
 import org.jetbrains.kotlin.psi.psiUtil.getSuperNames
 import java.lang.reflect.Modifier
+import kotlin.reflect.KClass
 
+@Suppress("TooManyFunctions")
```
Comment (chao2zhang) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleVisitor.kt:23]: Let's move the methods of `KtClassOrObject.requiresTypeResolution()` and `KtClassOrObject.isAnnotatedWith()` to `detekt-api`. I think they should be exposed and can be reused.

Hopefully we won't require suppressing `TooManyFunctions`

**Review by chao2zhang — APPROVED** [review]

**Comment by marschwar** [comment]
I removed the annotation from the exposed api by moving it to the internal package. Also the retention is changed to SOURCE so it is not available at runtime. This way we should be fairly flexible when it comes to changing the annotation(s) without breaking peoples code.

**Review by cortinico — APPROVED** [review]
> I still have some formatting issues. That is why there is one commit that just reformats the code according to the ktlint plugin

I wonder why this is happening 🤔 

> I started out adding @UnstableApi to the annotation but that would all rules to require to add @OptIn(UnstableApi::class) which does not feel right either.

I think `@UnstableApi` is a bit of an overkill. Having the annotation inside `.internal` is sufficient. Also I don't know how many users will start relying on this annotation right after, so I won't worry too much about protecting this API.

**Review by chao2zhang — APPROVED** [review]
@cortinico I was wondering if this should be in external API. The problem is that we should also allow custom rules to be annotated correctly, #3577 is a good example to fully skip type resolution rules.

Regardless, this PR looks good as it stands for now. I don't mind if we "escalate" these apis later.

File: detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt [hunk 2]
```diff
@@ -203,6 +203,32 @@ class RuleCollectorSpec : Spek({
...
+                /**
+                 * description
+                 */
+                @io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
+                class SomeRandomClass : Rule
+            """
+            val items = subject.run(code)
+            assertThat(items[0].requiresTypeResolution).isTrue()
+        }
+
```
Comment (BraisGabin) [inline detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt:231]: Are we lacking a test to check that this value is false if we don't have the annotation?

**Review by BraisGabin — APPROVED** [review]

**Comment by cortinico** [comment]
> @cortinico I was wondering if this should be in external API. The problem is that we should also allow custom rules to be annotated correctly

I believe yes, this should be an external API. Something to remember for the future 👍

**Review by chao2zhang — APPROVED** [review]

File: detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt:231]: I can follow up with the test. This PR has been open for a while so I would merge this in as we are adding more annotations.

**Pull request merged**